druntime/src/core/internal/gc/impl/conservative/gc.d

```diff
@@ -3759,29 +3730,62 @@ Lmark:
         }
 
         ScanRange!precise rng;
-        alias toscan = scanStack!precise;
 
-        while (atomicLoad(busyThreads) > 0)
+        stackLock.lock();
+        while (!scanStack.empty())
         {
-            if (toscan.empty)
-            {
-                evDone.wait(dur!"msecs"(1));
-                continue;
-            }
-
             busyThreads.atomicOp!"+="(1);
-            if (toscan.popLocked(rng))
+            scanStack.pop(rng);
+            if (scanStack.empty)
+                evStackFilled.reset();
+            stackLock.unlock();
+
+            version (Posix) debug (PARALLEL_PRINTF)
             {
-                version (Posix) debug (PARALLEL_PRINTF)
-                {
-                    printf("scanBackground thread %d scanning range [%p,%lld] from stack\n",
-                        threadId, rng.pbot, cast(long) (rng.ptop - rng.pbot));
-                }
-                mark!(precise, true, true)(rng);
+                printf("scanBackground thread %d scanning range [%p,%lld] from stack\n",
+                    threadId, rng.pbot, cast(long) (rng.ptop - rng.pbot));
             }
+            mark!(precise, true, true)(rng);
+            // returns here only if an empty scan stack has been seen
+
+            stackLock.lock();
             busyThreads.atomicOp!"-="(1);
         }
+        bool cont = busyThreads > 0;
+        stackLock.unlock();
+
         version (Posix) debug (PARALLEL_PRINTF) printf("scanBackground thread %d done\n", threadId);
+        return cont;
+    }
+
+    auto stackLock = shared(AlignedSpinLock)(SpinLock.Contention.brief);
+
+    void scanStackPushLocked(RANGE)(RANGE[] ranges)
+    {
+        stackLock.lock();
+        scope(exit) stackLock.unlock();
+        bool wasEmpty = scanStack.empty;
+
+        // reverse order for depth-first-order traversal
+        foreach_reverse (ref range; ranges)
+            scanStack.push(range);
```

Comment:
I know it's not a huge deal here, but this does a lot of checking if the stack needs expanding when we know the size we will require.

Reply:
I was also considering adding a primitive adding multiple ranges to avoid the duplicate "logic" of the foreach_reverse-loop for the parallel and single-threaded case. Avoiding the checks is another reason, so I added `pushReverse`.